### PR TITLE
AP_Mount, AP_Camera: Load CAMERA_INFORMATION from JSON

### PR DESCRIPTION
**Pull request by nexton-winjeel**
This PR allows us to add a `*.json` file to the SD card (or ROMFS) that is used to populate the [`CAMERA_INFORMATION`](https://mavlink.io/en/messages/common.html#CAMERA_INFORMATION) and [`VIDEO_STREAM_INFORMATION`](https://mavlink.io/en/messages/common.html#VIDEO_STREAM_INFORMATION) MAVLink messages. These messages are sent for `AP_Mount` backends that do not already implement these messages.

Sending these messages allows a GCS to auto-configure the video stream to receive video from the payload.

### CODE REVIEW

**Comment by rmackay9** [comment]
really great to see this!

**Comment by nexton-winjeel** [comment]
Two issues:
1. the test vectors in the new tests fail the formatter; and
2. the `shutil.copy()` calls in the new tests fail if the intermediate directories aren't present. 

I'll fix them now...

**Comment by nexton-winjeel** [comment]
@rmackay9: I'll jump on the dev call tomorrow if you want to discuss this.

File: Tools/autotest/arducopter.py [hunk 1]
```diff
@@ -11559,6 +11559,115 @@ def GripperReleaseOnThrustLoss(self):
         self.context_pop()
         self.reboot_sitl()
 
+    def CameraInformationFromJSON(self):
+        '''Tests that we can send CAMERA_INFORMATION message defined in JSON'''
+
+        json_filename = "camera_information.json"
+
+        # We use an example JSON file for this test...
+        example_json = os.path.realpath(os.path.join(testdir, "../../libraries/AP_Camera/json/", json_filename))
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py:11568]: with no try/except block around this stuff we may end up leaving stuff lying around that we shouldn't.

We fix this with scripting with `install_test_script_context` and friends.

I wonder if we need to generalise that stuff allow copying/deletion of arbitrary files.

File: libraries/AP_Camera/AP_Camera_Backend.cpp [hunk 2]
```diff
@@ -5,6 +5,16 @@
...
 #include <AP_Mount/AP_Mount.h>
 
+#if 0
+    #define debug(fmt, args ...) do { hal.console->printf("%s:%d: " fmt "\n", __FUNCTION__, __LINE__, ## args); } while(0)
+#else
+    #define debug(fmt, args ...)
+#endif
+
+#if AP_CAMERA_JSON_INFO_ENABLED
+    #include <AP_Filesystem/AP_Filesystem.h>
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera_Backend.cpp:15]: We don't usually indent these lines

File: libraries/AP_Camera/AP_Camera_Backend.cpp [hunk 3]
```diff
@@ -415,4 +449,178 @@ void AP_Camera_Backend::log_picture()
 }
 #endif
 
+#if AP_CAMERA_JSON_INFO_ENABLED
+// Read a JSON file from the expected folder on the SD card. If not found, will
+// fall back to looking for the same folder in ROMFS.
+//
+// This allocates a AP_JSON::value object that will need to be freed.
+AP_JSON::value * AP_Camera_Backend::_load_mount_msg_json(const char* json_filename) {
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera_Backend.cpp:457]:
```suggestion
AP_JSON::value * AP_Camera_Backend::_load_mount_msg_json(const char* json_filename)
{
```

similarly elsewhere

File: libraries/AP_Camera/AP_Camera_Backend.cpp [hunk 4]
```diff
@@ -415,4 +449,178 @@ void AP_Camera_Backend::log_picture()
...
+    char* romfs_json_path = nullptr;
+    int alloc = asprintf(&romfs_json_path, "@ROMFS/mav_msg_def/AP_Camera/%s", json_filename);
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera_Backend.cpp:459]: Do without the local here?

File: libraries/AP_Camera/AP_Camera_Backend.cpp [hunk 5]
```diff
@@ -415,4 +449,178 @@ void AP_Camera_Backend::log_picture()
...
+            debug("failed to parse '%s'", json_path);
+        }
+    }
+
+    free(romfs_json_path);
+    return obj;
+}
+
+// helper function to copy a JSON double into a msg struct
+template <typename T> bool AP_Camera_Backend::_copy_json_field_double(const AP_JSON::value* obj, const char* key, T& dst) {
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera_Backend.cpp:492]:
```suggestion
template <typename T>
bool AP_Camera_Backend::_copy_json_field_double(const AP_JSON::value* obj, const char* key, T& dst) {
```

File: libraries/AP_Camera/AP_Camera_Backend.cpp [hunk 6]
```diff
@@ -415,4 +449,178 @@ void AP_Camera_Backend::log_picture()
...
+    if (!_copy_json_field_double(obj, "resolution_v", camera_info.msg.resolution_v)) goto err;
+    if (!_copy_json_field_double(obj, "flags", camera_info.msg.flags)) goto err;
+    if (!_copy_json_field_double(obj, "cam_definition_version", camera_info.msg.cam_definition_version)) goto err;
+
+    if (!_copy_json_field_string(obj, "vendor_name", (char*)camera_info.msg.vendor_name, sizeof(camera_info.msg.vendor_name))) goto err;
+    if (!_copy_json_field_string(obj, "model_name", (char*)camera_info.msg.model_name, sizeof(camera_info.msg.model_name))) goto err;
+    if (!_copy_json_field_string(obj, "cam_definition_uri", camera_info.msg.cam_definition_uri, sizeof(camera_info.msg.cam_definition_uri))) goto err;
+
+    // Populate the fields that shouldn't change.
+    camera_info.msg.time_boot_ms = AP_HAL::millis();
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera_Backend.cpp:589]: Errr.. this one should definitely change

File: libraries/AP_Camera/AP_Camera_Backend.cpp [hunk 7]
```diff
@@ -415,4 +449,178 @@ void AP_Camera_Backend::log_picture()
...
+        debug("    camera_info.msg.flags=%u", camera_info.msg.flags);
+        debug("    camera_info.msg.cam_definition_version=%u", camera_info.msg.cam_definition_version);
+        debug("    camera_info.msg.cam_definition_uri='%s'", camera_info.msg.cam_definition_uri);
+        debug("    camera_info.msg.gimbal_device_id=%u", camera_info.msg.gimbal_device_id);
+    }
+}
+
+#undef GET_JSON_DOUBLE
+#undef GET_JSON_STR
+
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera_Backend.cpp:622]: Were these replaced by the template function?

File: libraries/AP_Camera/AP_Camera_Mount.cpp [hunk 8]
```diff
@@ -87,7 +87,16 @@ void AP_Camera_Mount::send_camera_information(mavlink_channel_t chan) const
 {
     AP_Mount* mount = AP::mount();
     if (mount != nullptr) {
-        return mount->send_camera_information(get_mount_instance(), chan);
+        bool sent_by_mount = mount->send_camera_information(get_mount_instance(), chan);
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera_Mount.cpp:90]:
```suggestion
        const bool sent_by_mount = mount->send_camera_information(get_mount_instance(), chan);
```

File: libraries/AP_Camera/AP_Camera_Mount.cpp [hunk 9]
```diff
@@ -109,4 +118,22 @@ void AP_Camera_Mount::send_camera_capture_status(mavlink_channel_t chan) const
     }
 }
 
+// send video stream information message to GCS
+void AP_Camera_Mount::send_video_stream_information(mavlink_channel_t chan) const
+{
+    AP_Mount* mount = AP::mount();
+    if (mount != nullptr) {
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera_Mount.cpp:125]:
```suggestion
    AP_Mount* mount = AP::mount();
    if (mount == nullptr) {
        return;
    }
```

File: libraries/AP_Camera/json/video_stream_information.json [hunk 10]
```diff
@@ -0,0 +1,20 @@
+# This file is used to populate the VIDEO_STREAM_INFORMATION message sent by
```
Comment (peterbarker) [inline libraries/AP_Camera/json/video_stream_information.json:1]: SHould we put these under an "examples" subdirectory as we do for a lot of other things?

**Review by peterbarker — CHANGES_REQUESTED** [review]
We should make sure that when compiled out these new features take up approximately no space.

Won't be able to be no space given different return values for some methods now

File: libraries/AP_Camera/AP_Camera.cpp [hunk 11]
```diff
@@ -462,6 +462,10 @@ bool AP_Camera::send_mavlink_message(GCS_MAVLINK &link, const enum ap_message ms
         CHECK_PAYLOAD_SIZE2(CAMERA_CAPTURE_STATUS);
         send_camera_capture_status(chan);
         break;
+    case MSG_VIDEO_STREAM_INFO:
```
[the remark this answers is not in the document]
Comment (nexton-winjeel) [reply libraries/AP_Camera/AP_Camera.cpp:466]: Done.

File: libraries/AP_Camera/AP_Camera_Backend.cpp (hunk 2 above)
Comment (nexton-winjeel) [reply libraries/AP_Camera/AP_Camera_Backend.cpp]: Fixed.

File: libraries/AP_Camera/AP_Camera_Backend.cpp (hunk 3 above)
Comment (nexton-winjeel) [reply libraries/AP_Camera/AP_Camera_Backend.cpp]: Fixed.

File: libraries/AP_Camera/AP_Camera_Backend.cpp (hunk 4 above)
Comment (nexton-winjeel) [reply libraries/AP_Camera/AP_Camera_Backend.cpp:466]: I find it more readable with the local.

File: libraries/AP_Camera/AP_Camera_Backend.cpp (hunk 5 above)
Comment (nexton-winjeel) [reply libraries/AP_Camera/AP_Camera_Backend.cpp]: Fixed.

File: libraries/AP_Camera/AP_Camera_Backend.cpp (hunk 6 above)
Comment (nexton-winjeel) [reply libraries/AP_Camera/AP_Camera_Backend.cpp]: Fixed.

File: libraries/AP_Camera/AP_Camera_Backend.cpp (hunk 7 above)
Comment (nexton-winjeel) [reply libraries/AP_Camera/AP_Camera_Backend.cpp]: Yes, they were. Fixed.

File: libraries/AP_Camera/AP_Camera_Mount.cpp (hunk 8 above)
Comment (nexton-winjeel) [reply libraries/AP_Camera/AP_Camera_Mount.cpp]: Fixed.

File: libraries/AP_Camera/AP_Camera_Mount.cpp (hunk 9 above)
Comment (nexton-winjeel) [reply libraries/AP_Camera/AP_Camera_Mount.cpp]: Fixed.

File: libraries/AP_Camera/json/video_stream_information.json (hunk 10 above)
Comment (nexton-winjeel) [reply libraries/AP_Camera/json/video_stream_information.json:1]: I'd prefer they live in the `AP_Camera` folder. The code in the backend makes assumptions about what fields are and aren't present in the JSON.

**Comment by nexton-winjeel** [comment]
Temporarily back to draft -- I need to confirm that this still works on hardware.

**Comment by nexton-winjeel** [comment]
Running on a CubeOrange with debug enabled:
```
STABILIZE> reboot
STABILIZE> Got COMMAND_ACK: PREFLIGHT_REBOOT_SHUTDOWN: ACCEPTED
Device /dev/serial/by-id/usb-Hex_ProfiCNC_CubeOrange_24004E001351313139383237-if00 is dead
Device /dev/serial/by-id/usb-Hex_ProfiCNC_CubeOrange_24004E001351313139383237-if00 reopened OK
AP_Logger_File: buffer size=204800
MS5611 found on bus 4 address 0x02
MS5611 found on bus 1 address 0x03
AP: Initialising ArduPilot
disabling flow control on serial 1
_load_mount_msg_json:479: failed to load 'mav_msg_def/AP_Camera/camera_information.json'
init_camera_information_from_json:616: Loaded camera_info from 'camera_information.json'
init_camera_information_from_json:617:     camera_info.msg.time_boot_ms=0
init_camera_information_from_json:618:     camera_info.msg.vendor_name='Unknown'
init_camera_information_from_json:619:     camera_info.msg.model_name='Camera'
init_camera_information_from_json:620:     camera_info.msg.firmware_version=0
init_camera_information_from_json:621:     camera_info.msg.focal_length=1.600000
init_camera_information_from_json:622:     camera_info.msg.sensor_size_h=3840.000
init_camera_information_from_json:623:     camera_info.msg.sensor_size_v=2160.000
init_camera_information_from_json:624:     camera_info.msg.resolution_h=1920
init_camera_information_from_json:625:     camera_info.msg.resolution_v=1080
init_camera_information_from_json:626:     camera_info.msg.lens_id=1
init_camera_information_from_json:627:     camera_info.msg.flags=256
init_camera_information_from_json:628:     camera_info.msg.cam_definition_version=0
init_camera_information_from_json:629:     camera_info.msg.cam_definition_uri=''
init_camera_information_from_json:630:     camera_info.msg.gimbal_device_id=1
_load_mount_msg_json:479: failed to load 'mav_msg_def/AP_Camera/video_stream_information.json'
init_video_stream_information_from_json:560: Loaded video_stream_info from 'video_stream_information.json'
init_video_stream_information_from_json:561:     video_stream_info.msg.stream_id=1
init_video_stream_information_from_json:562:     video_stream_info.msg.count=1
init_video_stream_information_from_json:563:     video_stream_info.msg.type=1
init_video_stream_information_from_json:564:     video_stream_info.msg.flags=1
init_video_stream_information_from_json:565:     video_stream_info.msg.framerate=30.00000
init_video_stream_information_from_json:566:     video_stream_info.msg.resolution_h=1920
init_video_stream_information_from_json:567:     video_stream_info.msg.resolution_v=1080
init_video_stream_information_from_json:568:     video_stream_info.msg.bitrate=1500
init_video_stream_information_from_json:569:     video_stream_info.msg.rotation=0
init_video_stream_information_from_json:570:     video_stream_info.msg.hfov=50
init_video_stream_information_from_json:572:     video_stream_info.msg.name='Video'
init_video_stream_information_from_json:573:     video_stream_info.msg.uri='127.0.0.1:5600'
AP: Calibrating barometer
AP: Barometer 1 calibration complete
AP: Barometer 2 calibration complete
Init Gyro**
AP: ArduPilot Ready
```

**Comment by nexton-winjeel** [comment]
@rmackay9, @peterbarker: Closing in favour of https://github.com/ArduPilot/ardupilot/pull/27794

**Pull request closed without merging**